drv/gimlet-hf-server/src/main.rs

```diff
@@ -666,75 +640,48 @@ impl idl::InOrderHostFlashImpl for ServerImpl {
     ) -> Result<(), RequestError<HfError>> {
         let data = HfPersistentData { dev_select };
         self.check_muxed_to_sp()?;
-        if self.dev_select_pin.is_some() {
-            let prev_slot = self.dev_state;
-
-            // At this point, all of the ways that `set_dev` could fail have
-            // been checked: we already called both `check_muxed_to_sp` and that
-            // `dev_select_pin` is `Some(...)`, so we can unwrap returns from
-            // `self.set_dev(...)`.
-            self.set_dev(!prev_slot).unwrap();
-            let (a_data, a_next) = self.persistent_data_scan();
-
-            self.set_dev(prev_slot).unwrap();
-            let (b_data, b_next) = self.persistent_data_scan();
-
-            let prev_monotonic_counter = match (a_data, b_data) {
-                (Some(a), Some(b)) => {
-                    a.monotonic_counter.max(b.monotonic_counter)
-                }
-                (Some(a), None) => a.monotonic_counter,
-                (None, Some(b)) => b.monotonic_counter,
-                (None, None) => 0,
-            };
+        let prev_slot = self.dev_state;
 
-            // Early exit if the previous persistent data matches
-            let prev_raw =
-                HfRawPersistentData::new(data, prev_monotonic_counter);
-            if a_data == b_data && a_data == Some(prev_raw) {
-                return Ok(());
-            }
+        // After having called `check_muxed_to_sp`, `self.set_dev(..)` is
+        // infallible and we can unwrap its returns.
+        self.set_dev(!prev_slot).unwrap();
+        let (a_data, a_next) = self.persistent_data_scan();
 
-            let monotonic_counter = prev_monotonic_counter
-                .checked_add(1)
-                .ok_or(HfError::MonotonicCounterOverflow)?;
-            let raw = HfRawPersistentData::new(data, monotonic_counter);
-
-            // Write the persistent data to the currently inactive flash.
-            self.set_dev(!prev_slot).unwrap();
-            let out_a = self.write_raw_persistent_data_to_addr(a_next, &raw);
-
-            // Swap back to the currently selected flash
-            self.set_dev(prev_slot).unwrap();
-
-            // Now that we've restored the current active flash, check whether
-            // we should propagate errors.
-            out_a?;
-
-            // Write the persistent data to the currently active flash
-            self.write_raw_persistent_data_to_addr(b_next, &raw)?;
-        } else {
-            // Single-flash case is less complicated, because we don't have to
-            // track and maintain the active device.
-            let (prev_data, next) = self.persistent_data_scan();
-            let prev_monotonic_counter = match prev_data {
-                Some(a) => a.monotonic_counter,
-                None => 0,
-            };
+        self.set_dev(prev_slot).unwrap();
+        let (b_data, b_next) = self.persistent_data_scan();
```

Comment:
Doesn't have to be here but do we actually need  the `unwraps` here? We also call `check_muxed_to_sp` in `set_dev` too.

Reply:
Good catch, we could either discard the value (`let _ = self.set_dev(...);`) or do normal error propagation without changing the resulting behavior.  I'm going to leave the unwraps here to minimize changes, but we could revisit in the future; the downside would be that we'd lose an implicit assertion about expected behavior.